src/main/java/com/iota/iri/Snapshot.java

```diff
@@ -2,175 +2,177 @@
-        return new Snapshot(state, index);
+    public Snapshot copySnapshot() {
+        rwlock.readLock().lock();
+        try {
+            return new Snapshot(state, index);
+        } finally {
+            rwlock.readLock().unlock();
+        }
     }
 
     public Long getBalance(Hash hash) {
-        Long l;
         rwlock.readLock().lock();
-        l = state.get(hash);
-        rwlock.readLock().unlock();
-        return l;
+        try {
+            return state.get(hash);
+        } finally {
+            rwlock.readLock().unlock();
+        }
     }
 
     public Map<Hash, Long> patchedDiff(Map<Hash, Long> diff) {
-        Map<Hash, Long> patch;
         rwlock.readLock().lock();
-        patch = diff.entrySet().stream().map(hashLongEntry ->
-            new HashMap.SimpleEntry<>(hashLongEntry.getKey(), state.getOrDefault(hashLongEntry.getKey(), 0L) + hashLongEntry.getValue())
-        ).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
-        rwlock.readLock().unlock();
-        return patch;
+        try {
+            Map<Hash, Long> map = new HashMap<>(diff.size());
+            diff.forEach((key, value) -> {
+                Long previousValue = state.getOrDefault(key, 0L);
+                map.put(key, Math.addExact(previousValue, value));
+            });
+            return map;
+        } finally {
+            rwlock.readLock().unlock();
+        }
     }
 
     void apply(Map<Hash, Long> patch, int newIndex) {
         if (!patch.entrySet().stream().map(Map.Entry::getValue).reduce(Math::addExact).orElse(0L).equals(0L)) {
             throw new RuntimeException("Diff is not consistent.");
         }
         rwlock.writeLock().lock();
-        patch.entrySet().stream().forEach(hashLongEntry -> {
-            if (state.computeIfPresent(hashLongEntry.getKey(), (hash, aLong) -> hashLongEntry.getValue() + aLong) == null) {
-                state.putIfAbsent(hashLongEntry.getKey(), hashLongEntry.getValue());
-            }
-        });
-        index = newIndex;
-        rwlock.writeLock().unlock();
+        try {
+            patch.forEach((key, value) -> {
+                Long previousValue = state.getOrDefault(key, 0L);
+                state.put(key, Math.addExact(value, previousValue));
+            });
+            index = newIndex;
+        } finally {
+            rwlock.writeLock().unlock();
+        }
     }
 
     public static boolean isConsistent(Map<Hash, Long> state) {
-        final Iterator<Map.Entry<Hash, Long>> stateIterator = state.entrySet().iterator();
-        while (stateIterator.hasNext()) {
-
-            final Map.Entry<Hash, Long> entry = stateIterator.next();
-            if (entry.getValue() <= 0) {
-
-                if (entry.getValue() < 0) {
-                    log.info("Skipping negative value for address: " + entry.getKey() + ": " + entry.getValue());
-                    return false;
-                }
-
-                stateIterator.remove();
```

Comment:
In all instances, this method is called on maps that are never used in the end. So the call to remove does nothing. The only map it could apply to have any effect on is the initial snapshot. It has been eliminated and added into init().